Update SUNDIALS support to version 2.6

for the record, since version 4.1.0 of Sundials (Feb 2019): "The implementation header file cvode impl.h is no longer installed. This means users who are directly manipulating the CVodeMem structure will need to update their code to use cvode’s public API."

This breaks compilation:

```
[  1%] Building C object CMakeFiles/partmclib.dir/src/condense_solver.c.o
/data/keeling/a/sarabas/devel/partmc/src/condense_solver.c:15:10: fatal error: cvode/cvode_impl.h: No such file or directory
   15 | #include <cvode/cvode_impl.h>
      |          ^~~~~~~~~~~~~~~~~~~~
```

Related FEniCS issues:
- https://bitbucket.org/fenics-project/dolfin/issues/1067/cv_functional-cv_newton-cv_adams-not
- https://aur.archlinux.org/packages/dolfin/?O=10&PP=10

--- a/.github/workflows/cloud_parcel.yml
+++ b/.github/workflows/cloud_parcel.yml
@@ -1,0 +1,15 @@
+name:  cloud-parcel
+
+on: [push, pull_request]
+
+jobs:
+  build:
+
+    runs-on: ubuntu-latest
+
+    steps:
+    - uses: actions/checkout@v1
+    - name: Build the testing docker image
+      run: docker build -f Dockerfile.cloud_parcel -t partmc-cloud-parcel .
+    - name: Run the tests in the docker container
+      run: docker run -t partmc-cloud-parcel bash -c 'cd /build; make test'

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -103,30 +103,6 @@ endif()
 # SUNDIALS
 
 if(ENABLE_SUNDIALS)
-  find_path(SUITE_SPARSE_INCLUDE_DIR klu.h
-    DOC "SuiteSparse include directory (must have klu.h)"
-    PATHS $ENV{SUITE_SPARSE_HOME}/include $ENV{SUNDIALS_HOME}/include
-          /opt/local/include /usr/local/include)
-  find_library(SUITE_SPARSE_KLU_LIB klu
-    DOC "SuiteSparse klu library"
-    PATHS $ENV{SUITE_SPARSE_HOME}/lib $ENV{SUNDIALS_HOME}/lib
-          /opt/local/lib /usr/local/lib)
-  find_library(SUITE_SPARSE_AMD_LIB amd
-    DOC "SuiteSparse amd library"
-    PATHS $ENV{SUITE_SPARSE_HOME}/lib $ENV{SUNDIALS_HOME}/lib
-          /opt/local/lib /usr/local/lib)
-  find_library(SUITE_SPARSE_BTF_LIB btf
-    DOC "SuiteSparse btf library"
-    PATHS $ENV{SUITE_SPARSE_HOME}/lib $ENV{SUNDIALS_HOME}/lib
-          /opt/local/lib /usr/local/lib)
-  find_library(SUITE_SPARSE_COLAMD_LIB colamd
-    DOC "SuiteSparse colamd library"
-    PATHS $ENV{SUITE_SPARSE_HOME}/lib $ENV{SUNDIALS_HOME}/lib
-          /opt/local/lib /usr/local/lib)
-  find_library(SUITE_SPARSE_CONFIG_LIB suitesparseconfig
-    DOC "SuiteSparse config library"
-    PATHS $ENV{SUITE_SPARSE_HOME}/lib $ENV{SUNDIALS_HOME}/lib
-          /opt/local/lib /usr/local/lib)
   find_path(SUNDIALS_INCLUDE_DIR cvode/cvode.h
     DOC "SUNDIALS include directory (must have cvode/, sundials/, nvector/ subdirs)"
     PATHS $ENV{SUNDIALS_HOME}/include /opt/local/include /usr/local/include)
@@ -136,20 +112,9 @@ if(ENABLE_SUNDIALS)
   find_library(SUNDIALS_CVODE_LIB sundials_cvode
     DOC "SUNDIALS CVODE library"
     PATHS $ENV{SUNDIALS_HOME}/lib /opt/local/lib /usr/local/lib)
-  find_library(SUNDIALS_KLU_LIB sundials_sunlinsolklu
-    DOC "SUNDIALS KLU library"
-    PATHS $ENV{SUITE_SPARSE_HOME}/lib $ENV{SUNDIALS_HOME}/lib
-          /opt/local/lib /usr/local/lib)
-  find_library(SUNDIALS_SUNMATRIX_SPARSE_LIB sundials_sunmatrixsparse
-    DOC "SUNDIALS SUNMatrixSparse library"
-    PATHS $ENV{SUITE_SPARSE_HOME}/lib $ENV{SUNDIALS_HOME}/lib
-          /opt/local/lib /usr/local/lib)
-  set(SUNDIALS_LIBS ${SUNDIALS_NVECSERIAL_LIB} ${SUNDIALS_CVODE_LIB}
-    ${SUNDIALS_KLU_LIB} ${SUNDIALS_SUNMATRIX_SPARSE_LIB} ${SUITE_SPARSE_KLU_LIB}
-    ${SUITE_SPARSE_COLAMD_LIB} ${SUITE_SPARSE_AMD_LIB} ${SUITE_SPARSE_BTF_LIB}
-    ${SUITE_SPARSE_CONFIG_LIB})
+  set(SUNDIALS_LIBS ${SUNDIALS_NVECSERIAL_LIB} ${SUNDIALS_CVODE_LIB})
   set(SUNDIALS_SRC src/condense_solver.c)
-  include_directories(${SUNDIALS_INCLUDE_DIR} ${SUITE_SPARSE_INCLUDE_DIR})
+  include_directories(${SUNDIALS_INCLUDE_DIR})
   add_definitions(-DPMC_USE_SUNDIALS)
 endif()
 

--- a/Dockerfile
+++ b/Dockerfile
@@ -9,6 +9,7 @@ RUN dnf -y update \
         sundials-devel \
         openblas-devel \
         metis-devel \
+        lapack-devel \
         gsl-devel \
         cmake \
     && dnf clean all

--- a/Dockerfile
+++ b/Dockerfile
@@ -6,9 +6,18 @@ RUN dnf -y update \
         make \
         netcdf-fortran-devel \
         sundials-devel \
+        openblas-devel \
         gsl-devel \
         cmake \
     && dnf clean all
+
+# Build the SuiteSparse libraries for sparse matrix support
+# (-k included because of problem with SuiteSparse security certificate - 1 Aug 2021)
+RUN curl -kLO http://faculty.cse.tamu.edu/davis/SuiteSparse/SuiteSparse-5.1.0.tar.gz \
+    && tar -zxvf SuiteSparse-5.1.0.tar.gz \
+    && export CXX=/usr/bin/cc \
+    && cd SuiteSparse \
+    && make install INSTALL=/usr/local BLAS="-L/lib64 -lopenblas"
 
 # NOTE: Modify .dockerignore to whitelist files/directories to copy.
 COPY . /partmc/

--- a/Dockerfile
+++ b/Dockerfile
@@ -19,5 +19,6 @@ RUN mkdir /build \
              -D CMAKE_C_FLAGS_RELEASE="-O2 -g -Werror -Wall -Wextra" \
              -D CMAKE_Fortran_FLAGS_RELEASE="-O2 -g -Werror -fimplicit-none -Wall -Wextra -Wconversion -Wunderflow -Wimplicit-interface -Wno-compare-reals -Wno-unused -Wno-unused-parameter -Wno-unused-dummy-argument -fbounds-check" \
              -D ENABLE_GSL:BOOL=TRUE \
+             -D ENABLE_SUNDIALS:BOOL=TRUE \
              /partmc \
     && make

--- a/Dockerfile
+++ b/Dockerfile
@@ -7,6 +7,7 @@ RUN dnf -y update \
         netcdf-fortran-devel \
         sundials-devel \
         openblas-devel \
+        metis-devel \
         gsl-devel \
         cmake \
     && dnf clean all

--- a/Dockerfile
+++ b/Dockerfile
@@ -3,6 +3,7 @@ FROM fedora:33
 RUN dnf -y update \
     && dnf -y install \
         gcc-gfortran \
+        gcc-c++ \
         make \
         netcdf-fortran-devel \
         sundials-devel \

--- a/Dockerfile.cloud_parcel
+++ b/Dockerfile.cloud_parcel
@@ -5,6 +5,7 @@ RUN dnf -y update \
         gcc-gfortran \
         make \
         netcdf-fortran-devel \
+        sundials-devel \
         gsl-devel \
         cmake \
     && dnf clean all
@@ -18,5 +19,6 @@ RUN mkdir /build \
              -D CMAKE_C_FLAGS_RELEASE="-O2 -g -Werror -Wall -Wextra" \
              -D CMAKE_Fortran_FLAGS_RELEASE="-O2 -g -Werror -fimplicit-none -Wall -Wextra -Wconversion -Wunderflow -Wimplicit-interface -Wno-compare-reals -Wno-unused -Wno-unused-parameter -Wno-unused-dummy-argument -fbounds-check" \
              -D ENABLE_GSL:BOOL=TRUE \
+             -D ENABLE_SUNDIALS:BOOL=TRUE \
              /partmc \
     && make

--- a/README.md
+++ b/README.md
@@ -147,7 +147,7 @@ Optional dependencies:
    * MPI parallel support - <http://www.open-mpi.org/>
    * GSL for random number generators -
      <http://www.gnu.org/software/gsl/>
-   * SUNDIALS ODE solver version 2.6 for condensation support -
+   * SUNDIALS ODE solver for condensation support -
      <http://www.llnl.gov/casc/sundials/>
    * gnuplot for testcase plotting - <http://www.gnuplot.info/>
 

--- a/src/condense.F90
+++ b/src/condense.F90
@@ -22,7 +22,7 @@
 !!
 !! The SUNDIALS ODE solver is used to compute the system evolution
 !! using an implicit method. The system Jacobian is explicitly
-!! inverveted as its structure is very simple.
+!! inverted as its structure is very simple.
 !!
 !! All equations used in this file are written in detail in the file
 !! \c doc/condense.tex.

--- a/src/condense_solver.c
+++ b/src/condense_solver.c
@@ -60,6 +60,9 @@ static int condense_solver_Solve(double t, N_Vector ycur, N_Vector fcur,
 
 /*******************************************************/
 
+condense_vf_f(int neq, realtype t, double *y_f, double *ydot_f);
+condense_jac_solve_f(int neq, double t, double *ycur_f, double *fcur_f, double *b_f, double gamma);
+
 /** \brief Call the ODE solver.
  *
  * \param neq The number of equations.
@@ -154,7 +157,9 @@ int condense_solver(int neq, double *x_f, double *abstol_f, double reltol_f,
  * \param user_data A pointer to user-provided data.
  * \return A result code (0 is success).
  */
-static int condense_vf(realtype t, N_Vector y, N_Vector ydot, void*)
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wunused-parameter"
+static int condense_vf(realtype t, N_Vector y, N_Vector ydot, void *user_data)
 {
 	realtype *y_data, *ydot_data;
 	int i, neq;
@@ -179,6 +184,7 @@ static int condense_vf(realtype t, N_Vector y, N_Vector ydot, void*)
 	free(ydot_f);
 	return(0);
 }
+#pragma GCC diagnostic pop
 
 /** \brief Check the return value from a SUNDIALS call.
  *
@@ -234,8 +240,10 @@ static int condense_check_flag(void *flagvalue, char *funcname, int opt)
  * scalar, and \f$b\f$ is a given vector.
  *
  */
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wunused-parameter"
 static int condense_solver_Solve(double t, N_Vector ycur, N_Vector fcur,
-				 N_Vector b, N_Vector z, double gamma, double, int, void*)
+				 N_Vector b, N_Vector z, double gamma, double delta, int lr, void *user_data)
 {
 	realtype *b_data, *ycur_data, *fcur_data, *z_data;
 	int i, neq;
@@ -266,4 +274,4 @@ static int condense_solver_Solve(double t, N_Vector ycur, N_Vector fcur,
 	free(fcur_f);
 	return(0);
 }
-
+#pragma GCC diagnostic pop

--- a/src/condense_solver.c
+++ b/src/condense_solver.c
@@ -60,8 +60,8 @@ static int condense_solver_Solve(double t, N_Vector ycur, N_Vector fcur,
 
 /*******************************************************/
 
-condense_vf_f(int neq, realtype t, double *y_f, double *ydot_f);
-condense_jac_solve_f(int neq, double t, double *ycur_f, double *fcur_f, double *b_f, double gamma);
+void condense_vf_f(int neq, realtype t, double *y_f, double *ydot_f);
+void condense_jac_solve_f(int neq, double t, double *ycur_f, double *fcur_f, double *b_f, double gamma);
 
 /** \brief Call the ODE solver.
  *

--- a/src/condense_solver.c
+++ b/src/condense_solver.c
@@ -12,7 +12,7 @@
 #include <cvode/cvode.h>
 #include <nvector/nvector_serial.h>
 #include <sundials/sundials_types.h>
-#include <cvode/cvode_impl.h>
+#include <sunlinsol/sunlinsol_spgmr.h>
 
 /** \brief Result code indicating successful completion.
  */
@@ -38,6 +38,15 @@
 /** \brief Result code indicating failure of the solver.
  */
 #define PMC_CONDENSE_SOLVER_FAIL           7
+/** \brief Result code indicating failure in creation of the linear solver.
+ */
+#define PMC_CONDENSE_SOLVER_LINSOL_CTOR    8
+/** \brief Result code indicating failure in setting the linear solver.
+ */
+#define PMC_CONDENSE_SOLVER_LINSOL_SET     9
+/** \brief Result code indicating failure in setting the preconditioner.
+ */
+#define PMC_CONDENSE_SOLVER_LINSOL_PREC    10
 
 static int condense_vf(realtype t, N_Vector y, N_Vector ydot, void *user_data);
 
@@ -45,16 +54,10 @@ static int condense_check_flag(void *flagvalue, char *funcname, int opt);
 
 /*******************************************************/
 // solver block
-static int condense_solver_Init(CVodeMem cv_mem);
 
-static int condense_solver_Setup(CVodeMem cv_mem, int convfail, N_Vector ypred,
-				 N_Vector fpred, booleantype *jcurPtr,
-				 N_Vector vtemp1, N_Vector vtemp2, N_Vector vtemp3);
+static int condense_solver_Solve(double t, N_Vector ycur, N_Vector fcur,
+				 N_Vector r, N_Vector b, double gamma, double delta, int lr, void *user_data);
 
-static int condense_solver_Solve(CVodeMem cv_mem, N_Vector b, N_Vector weight,
-				 N_Vector ycur, N_Vector fcur);
-
-static void condense_solver_Free(CVodeMem cv_mem);
 /*******************************************************/
 
 /** \brief Call the ODE solver.
@@ -76,7 +79,6 @@ int condense_solver(int neq, double *x_f, double *abstol_f, double reltol_f,
 	realtype reltol, t_initial, t_final, t, tout;
 	N_Vector y, abstol;
 	void *cvode_mem;
-	CVodeMem cv_mem;
 	int flag, i, pretype, maxl;
 	realtype *y_data, *abstol_data;
 
@@ -102,7 +104,7 @@ int condense_solver(int neq, double *x_f, double *abstol_f, double reltol_f,
 	t_initial = t_initial_f;
 	t_final = t_final_f;
 
-	cvode_mem = CVodeCreate(CV_BDF, CV_NEWTON);
+	cvode_mem = CVodeCreate(CV_BDF);
 	if (condense_check_flag((void *)cvode_mem, "CVodeCreate", 0))
                 return PMC_CONDENSE_SOLVER_INIT_CVODE_MEM;
 
@@ -118,34 +120,16 @@ int condense_solver(int neq, double *x_f, double *abstol_f, double reltol_f,
 	if (condense_check_flag(&flag, "CVodeSetMaxNumSteps", 1))
                 return PMC_CONDENSE_SOLVER_SET_MAX_STEPS;
 
-        /*******************************************************/
-	// dense solver
-	//flag = CVDense(cvode_mem, neq);
-	//if (condense_check_flag(&flag, "CVDense", 1)) return(1);
-        /*******************************************************/
 
-        /*******************************************************/
-	// iterative solver
-	//pretype = PREC_LEFT;
-	//maxl = 0;
-	//flag = CVSptfqmr(cvode_mem, pretype, maxl);
-	//if (condense_check_flag(&flag, "CVSptfqmr", 1)) return(1);
-
-	//flag = CVSpilsSetJacTimesVecFn(cvode_mem, condense_jtimes);
-	//if (condense_check_flag(&flag, "CVSpilsSetJacTimesVecFn", 1)) return(1);
-
-	//flag = CVSpilsSetPreconditioner(cvode_mem, NULL, condense_prec);
-	//if (condense_check_flag(&flag, "CVSpilsSetPreconditioner", 1)) return(1);
-        /*******************************************************/
-
-        /*******************************************************/
-	// explicit solver
-	cv_mem = (CVodeMem)cvode_mem;
-	cv_mem->cv_linit = condense_solver_Init;
-	cv_mem->cv_lsetup = condense_solver_Setup;
-	cv_mem->cv_lsolve = condense_solver_Solve;
-	cv_mem->cv_lfree = condense_solver_Free;
-        /*******************************************************/
+        SUNLinearSolver LS = SUNLinSol_SPGMR(y, PREC_LEFT, 0);
+        if (condense_check_flag((void *)LS, "SUNLinSol_SPGMR", 0))
+                return PMC_CONDENSE_SOLVER_LINSOL_CTOR;
+	flag = CVodeSetLinearSolver(cvode_mem, LS, NULL);
+        if (condense_check_flag(&flag, "CVodeSetLinearSolver", 1))
+                return PMC_CONDENSE_SOLVER_LINSOL_SET;
+	flag = CVodeSetPreconditioner(cvode_mem, NULL, condense_solver_Solve);
+        if (condense_check_flag(&flag, "CVodeSetPreconditioner", 1))
+                return PMC_CONDENSE_SOLVER_LINSOL_PREC;
 
 	t = t_initial;
 	flag = CVode(cvode_mem, t_final, y, &t, CV_NORMAL);
@@ -243,67 +227,29 @@ static int condense_check_flag(void *flagvalue, char *funcname, int opt)
   return(0);
 }
 
-/** \brief Initialization routine for the ODE linear equation solver.
- *
- * \param cv_mem The \c CVODE solver parameter structure.
- * \return A result code (0 is success).
- */
-static int condense_solver_Init(CVodeMem cv_mem)
-{
-	return(0);
-}
-
-/** \brief Setup routine for the ODE linear equation solver.
- *
- * \param cv_mem The \c CVODE solver parameter structure.
- * \param convfail
- * \param ypred
- * \param fpred
- * \param jcurPtr
- * \param vtemp1
- * \param vtemp2
- * \param vtemp3
- * \return A result code (0 is success).
- */
-static int condense_solver_Setup(CVodeMem cv_mem, int convfail, N_Vector ypred,
-				 N_Vector fpred, booleantype *jcurPtr,
-				 N_Vector vtemp1, N_Vector vtemp2, N_Vector vtemp3)
-{
-	return(0);
-}
-
 /** \brief Linear solver routine for use by the ODE solver.
  *
  * Should solve the system \f$(I - \gamma J) x = b\f$, where \f$J\f$
  * is the current vector field Jacobian, \f$\gamma\f$ is a given
  * scalar, and \f$b\f$ is a given vector.
  *
- * \param cv_mem The \c CVODE solver parameter structure.
- * \param b The right-hand-side of the linear system.
- * \param weight
- * \param ycur The current state vector.
- * \param fcur The current vector field vector.
- * \return A result code (0 is success).
  */
-static int condense_solver_Solve(CVodeMem cv_mem, N_Vector b, N_Vector weight,
-				 N_Vector ycur, N_Vector fcur)
+static int condense_solver_Solve(double t, N_Vector ycur, N_Vector fcur,
+				 N_Vector b, N_Vector z, double gamma, double delta, int lr, void *user_data)
 {
-	realtype *b_data, *ycur_data, *fcur_data;
+	realtype *b_data, *ycur_data, *fcur_data, *z_data;
 	int i, neq;
 	double *b_f, *ycur_f, *fcur_f;
-	double t, gamma;
 
 	neq = NV_LENGTH_S(b);
 	b_data = NV_DATA_S(b);
+	z_data = NV_DATA_S(z);
 	ycur_data = NV_DATA_S(ycur);
 	fcur_data = NV_DATA_S(fcur);
 
 	b_f = malloc(neq * sizeof(double));
 	ycur_f = malloc(neq * sizeof(double));
 	fcur_f = malloc(neq * sizeof(double));
-
-	t = cv_mem->cv_tn;
-	gamma = cv_mem->cv_gamma;
 
 	for (i = 0; i < neq; i++) {
 		b_f[i] = b_data[i];
@@ -312,7 +258,7 @@ static int condense_solver_Solve(CVodeMem cv_mem, N_Vector b, N_Vector weight,
 	}
 	condense_jac_solve_f(neq, t, ycur_f, fcur_f, b_f, gamma);
 	for (i = 0; i < neq; i++) {
-		b_data[i] = b_f[i];
+		z_data[i] = b_f[i];
 	}
 
 	free(b_f);
@@ -321,10 +267,3 @@ static int condense_solver_Solve(CVodeMem cv_mem, N_Vector b, N_Vector weight,
 	return(0);
 }
 
-/** \brief Finalization routine for the ODE linear equation solver.
- *
- * \param cv_mem The \c CVODE solver parameter structure.
- */
-static void condense_solver_Free(CVodeMem cv_mem)
-{
-}

--- a/src/condense_solver.c
+++ b/src/condense_solver.c
@@ -76,10 +76,10 @@ static int condense_solver_Solve(double t, N_Vector ycur, N_Vector fcur,
 int condense_solver(int neq, double *x_f, double *abstol_f, double reltol_f,
 		    double t_initial_f, double t_final_f)
 {
-	realtype reltol, t_initial, t_final, t, tout;
+	realtype reltol, t_initial, t_final, t;
 	N_Vector y, abstol;
 	void *cvode_mem;
-	int flag, i, pretype, maxl;
+	int flag, i;
 	realtype *y_data, *abstol_data;
 
 	y = abstol = NULL;
@@ -154,7 +154,7 @@ int condense_solver(int neq, double *x_f, double *abstol_f, double reltol_f,
  * \param user_data A pointer to user-provided data.
  * \return A result code (0 is success).
  */
-static int condense_vf(realtype t, N_Vector y, N_Vector ydot, void *user_data)
+static int condense_vf(realtype t, N_Vector y, N_Vector ydot, void*)
 {
 	realtype *y_data, *ydot_data;
 	int i, neq;
@@ -235,7 +235,7 @@ static int condense_check_flag(void *flagvalue, char *funcname, int opt)
  *
  */
 static int condense_solver_Solve(double t, N_Vector ycur, N_Vector fcur,
-				 N_Vector b, N_Vector z, double gamma, double delta, int lr, void *user_data)
+				 N_Vector b, N_Vector z, double gamma, double, int, void*)
 {
 	realtype *b_data, *ycur_data, *fcur_data, *z_data;
 	int i, neq;

